soroban-react-dapp/contracts/greeting/src/lib.rs

```diff
@@ -1,24 +1,91 @@
 #![no_std]
-use soroban_sdk::{contract, contractimpl, Env, Symbol, symbol_short, String};
-
-const TITLE: Symbol = symbol_short!("TITLE");
-
-
+use soroban_sdk::{contract, contracterror, contractimpl, contracttype, Address, Env, String, Vec};
 #[contract]
 pub struct TitleContract;
 
+#[contracterror]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum Error {
+    Unauthorized = 1,
+}
+
 #[contractimpl]
 impl TitleContract {
+    pub fn init(env: Env, instantiator: Address) -> Result<(), Error> {
+        let storage = env.storage().instance();
+        instantiator.require_auth();
+        if storage.has(&DataKey::Instantiator) {
+            return Err(Error::Unauthorized);
+        }
+        storage.set(&DataKey::Instantiator, &instantiator);
+        Ok(())
+    }
 
-    pub fn set_title(env: Env, title: String) {
-                env.storage().instance().set(&TITLE, &title)
+    // sets the title in the storage
+    pub fn set_title(env: Env, caller: Address, title: String) -> Result<(), Error> {
+        caller.require_auth();
+        let storage = env.storage().instance();
+        let instantiator: Address = storage.get(&DataKey::Instantiator).unwrap();
+        let admins: Vec<Address> = storage.get(&DataKey::Admins).unwrap_or(Vec::new(&env));
+        if admins.contains(&caller) || caller.eq(&instantiator) {
+            env.storage().instance().set(&DataKey::Title, &title);
+            Ok(())
+        } else {
+            Err(Error::Unauthorized)
+        }
     }
 
+    // reads the title from the storage
     pub fn read_title(env: Env) -> String {
-        env.storage().instance().get(&TITLE)
+        env.storage()
+            .instance()
+            .get(&DataKey::Title)
             .unwrap_or(String::from_str(&env, "Default Title"))
     }
-    
-} 
+
+    // adds a new admin to the list of admins, only the instantiator can add an admin
+    pub fn add_admin(env: Env, new_admin: Address) {
+        let storage = env.storage().instance();
+        let instantiator: Address = storage.get(&DataKey::Instantiator).unwrap();
+        instantiator.require_auth();
+
+        let mut admins: Vec<Address> = storage.get(&DataKey::Admins).unwrap_or(Vec::new(&env));
+        if !admins.contains(&new_admin) {
```

Comment:
Vec stored in instance storage could lead to DoS. Not big deal since it's just the instantiator who is adding admins

Reply:
Ty! What would be a preferred approach?